fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -860,6 +873,88 @@ private ScoredPlan planCandidateScan(@Nonnull CandidateScan candidateScan,
         return null;
     }
 
+    @Nullable
+    private ScoredPlan optimizeCandidateScanForMultidimensionalIndex(@Nonnull PlanContext planContext,
+                                                                     @Nullable ScoredPlan p,
+                                                                     @Nullable final Index index) {
+        if (p == null) {
+            return null;
+        }
+        if (index == null || !index.getType().equals(IndexTypes.MULTIDIMENSIONAL)) {
+            return p;
+        }
+
+        // From now on, we may only discard the plan or make a better one
+
+        final RecordQueryIndexPlan oldIndexPlan = Objects.requireNonNull((RecordQueryIndexPlan)p.plan);
+        if (oldIndexPlan.isReverse()) {
+            return null;
+        }
+
+        final ComparisonRanges planComparisonRanges = getPlanComparisonRanges(p);
```

Comment:
The multidimensional and-with-then planner always guarantees to match each part of the index. The match may just be empty but it will be there in the data structure. There are checks on the prefix being equality-bound a little further down. I added a `Verify.verify()` that `planComparisonRanges` has the right arity.